mesh/mesh.hpp

```diff
@@ -2210,6 +2213,18 @@ class Mesh
        method does not modify the nodes. */
    void NodesUpdated() { DeleteGeometricFactors(); }
 
+   /// Returns the attributes for all elements in this mesh.
+   /// The returned array points to an internal object that may be invalidated
+   /// by mesh operations such as refinement or any element attributes are
+   /// modified. Since not all such modifications can be tracked by the Mesh
+   /// class (e.g. if a user calls GetElement() then changes the element
+   /// attribute directly), one needs to account for such changes by calling the
+   /// method ElementAttributesUpdated().
+   const Array<int>& GetElementAttributes() const;
+
+   /// Marks that element attribute data has been updated
+   void ElementAttributesUpdated();
```

Comment:
I think this is the purpose of `Mesh::SetAttributes`, we don't need a separate function for the attribute cache.

Also, the cache should be invalidated in `Mesh::ResetLazyData`.